File: 2017/elixir/day_02.exs
defmodule Day_02 do
  def star_01(input) do
    input
      |> Enum.map(&(Enum.max(&1) - Enum.min(&1)))
      |> Enum.sum
  end

  def compute_division_checksum([head | tail]) do
    checksum = compute_division_checksum(head, tail)
    if is_number(checksum) do
      checksum
    else
      compute_division_checksum(tail)
    end
  end

  defp compute_division_checksum(_a, []) do
    nil
  end

  defp compute_division_checksum(a, [a | tail]) do
    compute_division_checksum(a, tail)
  end

  defp compute_division_checksum(a, [b | tail]) do
    [min | [max]] = Enum.sort([a, b])
    if rem(max, min) == 0 do
      div(max, min)
    else
      compute_division_checksum(a, tail)
    end
  end

  def exercise_02(input) do
    input
      |> Enum.map(&compute_division_checksum/1)
      |> Enum.sum
  end
end

{:ok, file} = File.read "../inputs/day_02.txt"

parse_row = fn (row) ->
  row
    |> String.split(~r/\s+/)
    |> Enum.map(&String.to_integer/1)
end

input = file
  |> String.trim
  |> String.split("\n")
  |> Enum.map(parse_row)

IO.inspect Day_02.star_01(input)
IO.inspect Day_02.exercise_02(input)

File: 2017/javascript/day_02.js
const assert = require('assert')
const _ = require('redash')
const fs = require('fs')
const INPUT = fs.readFileSync('../inputs/day_02.txt', 'utf8')

/** Splits a multiline string into an array of rows */
const splitRows = _.pipe([
  _.split('\n'),
  _.reject(_.isEmpty),
])

/** Parses a single row (string) into an array of integers */
const parseRow = _.pipe([
  _.split(/\s+/),
  _.map(parseInt),
])

// Star #1
// -----------------------------------------------
const computeLeastGreatestChecksum = _.pipe([
  xs => xs.sort((a, b) => a - b),
  xs => _.last(xs) - _.head(xs),
])

const star_01 = _.pipe([
  splitRows,
  _.map(_.pipe([parseRow, computeLeastGreatestChecksum])),
  _.sum,
])

assert.equal(star_01(INPUT), 32020)

// Star #2
// -----------------------------------------------
const computeDivisibleChecksum = xs => {
  xs = xs.sort((a, b) => a - b)

  // O(n^2) :/
  for (let i = xs.length - 1; i > 0; i--) {
    for (let j = 0; j < i; j++) {
      const a = xs[i]
      const b = xs[j]
      if (a % b === 0) return a / b
    }
  }
}

const exercise_02 = _.pipe([
  splitRows,
  _.map(_.pipe([parseRow, computeDivisibleChecksum])),
  _.sum,
])

assert.equal(exercise_02(INPUT), 236)
